### PR TITLE
KAFKA-2686: Reset needsPartitionAssignment in SubscriptionState.assign()

### PR DESCRIPTION
**Pull request by guozhangwang**

### CODE REVIEW

**Comment by hachikuji** [comment]
Fix looks good to me. Should we add a test case in ConsumerTest?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 1]
```diff
@@ -129,11 +129,15 @@ public void groupSubscribe(Collection<String> topics) {
...
         this.groupSubscription.retainAll(subscription);
         this.needsPartitionAssignment = true;
     }
 
+    /**
+     * Change the assignment to the specified partitions provided by the user,
+     * note this is different from {@link #changePartitionAssignment(Collection)}
+     * whose input partitions are provided by the subscribed coordinator.
+     */
     public void assign(List<TopicPartition> partitions) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Would `userAssign` be a clearer name?

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 2]
```diff
@@ -306,6 +312,10 @@ public boolean partitionAssignmentNeeded() {
         return this.needsPartitionAssignment;
     }
 
+    /**
+     * Change the assignment to the specified partitions returned from the coordinator,
+     * note this is different from {@link #assign(List)} which directly set the assignment from user inputs
+     */
     public void changePartitionAssignment(Collection<TopicPartition> assignments) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Similarly, maybe we could call this `coordinatorAssign`?

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: I would prefer not involving coordinator in the subscription state class, maybe renaming to `assignFromSubscribed` and `assignFromUser`.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Not sure I see the harm in referring to the coordinator (we already do so in the comments), but that works too.

**Comment by guozhangwang** [comment]
Also changed the behavior of `subscribe(empty-list)` to be the same as `unsubscribe`.

**Comment by hachikuji** [comment]
LGTM

**Comment by guozhangwang** [comment]
Ping @junrao for review.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 3]
```diff
@@ -629,6 +629,9 @@ private KafkaConsumer(ConsumerConfig config,
      * assign partitions. Topic subscriptions are not incremental. This list will replace the current
      * assignment (if there is one). Note that it is not possible to combine topic subscription with group management
      * with manual partition assignment through {@link #assign(List)}.
+     *
+     * If the given list of topics is empty, then it treated the same as {@link #unsubscribe()}.
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: it treated => it is treated

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 4]
```diff
@@ -666,6 +674,9 @@ public void subscribe(List<String> topics, ConsumerRebalanceListener listener) {
      * assign partitions. Topic subscriptions are not incremental. This list will replace the current
      * assignment (if there is one). It is not possible to combine topic subscription with group management
      * with manual partition assignment through {@link #assign(List)}.
+     *
+     * If the given list of topics is empty, then it treated the same as {@link #unsubscribe()}.
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: it treated => it is treated

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 5]
```diff
@@ -129,12 +129,16 @@ public void groupSubscribe(Collection<String> topics) {
...
         this.needsPartitionAssignment = true;
     }
 
-    public void assign(List<TopicPartition> partitions) {
+    /**
+     * Change the assignment to the specified partitions provided by the user,
+     * note this is different from {@link #assignFromSubscribed(Collection)}
+     * whose input partitions are provided from the subscribed topics.
+     */
+    public void assignFromUser(List<TopicPartition> partitions) {
```
Comment (junrao) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: To be consistent with assignFromSubscribed(), should we take Collection<TopicPartition> in the input?

**Comment by junrao** [comment]
Thanks for the patch. LGTM. Just a few minor comments.

**Pull request closed without merging**